eslint/cactbot-output-strings.js

```diff
@@ -0,0 +1,111 @@
+        propKeys.push(prop.key.value);
+    } else if (t.isSpreadElement(prop)) {
+      if (t.isIdentifier(prop.argument) && spreadNameMaps.has(prop.argument.name)) {
+        getAllKeys(spreadNameMaps.get(prop.argument.name), spreadNameMaps)
+          .forEach((name) => propKeys.push(name));
+      }
+    }
+  });
+  return propKeys;
+};
+
+/**
+ * @type {import("eslint").Rule.RuleModule}
+ */
+const ruleModule = {
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'suggest outputStrings in cactbot',
+      category: 'Stylistic Issues',
+      recommended: true,
+      url: 'https://github.com/quisquous/cactbot/blob/main/docs/RaidbossGuide.md#trigger-elements',
+    },
+    fixable: 'code',
+    schema: [],
+    messages: {
+      noOutputStrings: 'no outputStrings in trigger',
+      notFoundProperty: 'not found \'{{prop}}\' in \'{{outputParam}}\'',
+    },
+  },
+  create: function(context) {
+    const sourceCode = context.getSourceCode();
+    const spreadNameMaps = new Map();
+    return {
+      VariableDeclarator(node) {
+        if (t.isIdentifier(node.id) && t.isObjectExpression(node.init))
+          spreadNameMaps.set(node.id.name, node.init.properties);
+      },
+      ObjectExpression(node) {
+        const propNames = getAllKeys(node.properties, spreadNameMaps);
+        if (propNames.includes('id') && propNames.some((name) => textProps.includes(name))) {
+          if (!node.properties.find((prop) => prop.key && prop.key.name === 'outputStrings')) {
+            context.report({
+              node: node,
+              messageId: 'noOutputStrings',
+            });
+          } else {
+            const outputStringProps = (() => {
+              const propNames = [];
+              const outputProps = node.properties.find((prop) => prop.key && prop.key.name === 'outputStrings');
+              if (outputProps && outputProps.value && outputProps.value.properties) {
+                getAllKeys(outputProps.value.properties, spreadNameMaps)
+                  .forEach((n) => propNames.push(n));
+              } else if (outputProps && t.isIdentifier(outputProps.value)) {
+                const obj = spreadNameMaps.get(outputProps.value.name);
+                if (obj)
+                  getAllKeys(obj, spreadNameMaps).forEach((n) => propNames.push(n));
+              }
+              return propNames;
+            })();
+            if (!outputStringProps) return;
+            const textFunctions = node
+              .properties
+              .filter((prop) => textProps.includes(prop.key.name))
+              .map((prop) => prop.value);
+            textFunctions.forEach((func) => {
+              if (t.isFunctionExpression(func) || t.isArrowFunctionExpression(func)) {
+                const param = func.params[2] && func.params[2].name;
+                const source = sourceCode.getText(func);
+                const m = source.match(new RegExp(`(?<=${param}\\.)(\\w+)(?=\\()`, 'g'));
```

Comment:
Can you `const param` => `const outputParam` for readability?

Do you need to check `if (!param) return;` here?

I think `(?<=${param}\\.)` should be `(?<=\b${param}\\.)` as well?

Reply:
Yes, yes.

And in fact, there is an issue here. In somewhere, there are something like `output[func(xxx)]()` to dynamically invoke outputs, and cannot be catching by this regex, which can only catch `output.prop(`.

Reply:
Yeah, I don't think there's anything to be done about using a variable.  Maybe typescript could help a little here, but it feels a bit unsolvable from a linting/testing perspective.